google/cloud/bigquery/_helpers.py

```diff
@@ -1004,3 +1005,61 @@ def _verify_job_config_type(job_config, expected_type, param_name="job_config"):
                 job_config=job_config,
             )
         )
+
+
+class ResourceBase:
+    """Base class providing the from_api_repr method."""
+
+    def __init__(self):
+        self._properties = {}
+
+    @classmethod
+    def from_api_repr(cls, resource: dict):
+        """Factory: constructs an instance of the class (cls)
+        given its API representation.
+
+        Args:
+            resource (Dict[str, Any]):
+                API representation of the object to be instantiated.
+
+        Returns:
+            ResourceBase: An instance of the class initialized with data
+                from 'resource'.
+        """
+        config = cls()
+        config._properties = copy.deepcopy(resource)
+        return config
+
+
+def _isinstance_or_raise(value: Any, dtype: Optional[Union[Any, Tuple[Any, ...]]]):
+    """Determine whether a value type matches a given datatype or None.
+
+    Args:
+        value (Any): Value to be checked.
+        dtype (Optional[Union[Any, Tuple[Any, ...]]]): Expected data type(s).
+
+    Returns:
+        Any: Returns the input value if the type check is successful.
+
+    Raises:
+        TypeError: If the input value's type does not match the expected data type(s).
+    """
+    
+    # Simplest case
+    if dtype is None and value is None:
+        return value
+    
+    elif isinstance(dtype, tuple):
+        # Another simple case
+        if None in dtype and value is None:
+            return value
+        # Iterate through the tuple and check if value is an instance of any type
+        if not any(isinstance(value, t) for t in dtype if t is not None):
+            valid_types_str = ", ".join(str(t) for t in dtype if t is not None)
+            msg = f"Pass {value} as one of '{valid_types_str}' or None. Got {type(value)}."
+            raise TypeError(msg)
+    else:
+        if not isinstance(value, dtype):
+            msg = f"Pass {value} as a '{dtype}'. Got {type(value)}."
+            raise TypeError(msg)
+    return value
```

Comment:
I think its a bit strange mixing None and types, but also the type hints not really being able to indicate that a type is expected. Since None is such a special case (as shown in the comment thread with Lingqing), I think it could make sense to split out the None part into its own argument, to effectively duplicate the isinstance() function signature, but add a boolean for if none is allowed.

```suggestion
def _isinstance_or_raise(value: Any, dtype: type, none_allowed: Optional[bool]):
    """Determine whether a value type matches a given datatype or None.

    Args:
        value (Any): Value to be checked.
        dtype (type): Expected data type(s).
        none_allowed Optional(bool): whether value is allowed to be None.

    Returns:
        Any: Returns the input value if the type check is successful.

    Raises:
        TypeError: If the input value's type does not match the expected data type(s).
    """
    if none_allowed and value is None:
        return value

    if isinstance(value, dtype):
        return value
    
    msg = f"Pass {value} as a '{dtype}' (or None). Got {type(value)}." # Add the 'or None' conditionally
    raise TypeError(msg)
```

WDYT?